### PR TITLE
Make new type param suggestion more targetted

### LINKED ISSUE #72640: Do not suggest type parameter when followed by `<`
```rust
struct S {
    m: Vec<Hashmap<String, ()>>,
}
```

```console
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:2:12
  |
1 | struct S {
  |         - help: you might be missing a type parameter: `<Hashmap>`
2 |     m: Vec<Hashmap<String, ()>>,
  |            ^^^^^^^ not found in this scope
```

Rustc should know that `struct S<Hashmap>` is not in the right direction because that still wouldn't make `Hashmap<String, ()>` a valid thing to write; type parameters can't have type parameters (yet??).

Separately, #70572 should also have prevented the suggestion in this case, but is an independent issue.

cc @estebank who worked on this recently in #68447.

### PR DESCRIPTION
**Pull request by estebank**
Do not suggest new type param when encountering a missing type in an ADT
field with generic parameters.

Fix #72640.

### CODE REVIEW

File: src/test/ui/associated-types/associated-types-eq-1.stderr [hunk 1]
```diff
@@ -4,7 +4,16 @@ error[E0412]: cannot find type `A` in this scope
...
+   |            ^
+   |
+help: a type parameter with a similar name exists
+   |
+LL |     let _: I = x.boo();
+   |            ^
+help: you might be missing a type parameter
+   |
+LL | fn foo2<I: Foo, A>(x: I) {
+   |               ^^^
```
Comment (estebank) [inline src/test/ui/associated-types/associated-types-eq-1.stderr:16]: Opened https://github.com/rust-lang/rust/issues/73321 for handling this more gracefully.

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -225,13 +225,15 @@ enum VisResolutionError<'a> {
...
-// `PathSegment`.
+/// A minimal representation of a path segment. We use this in resolve because we synthesize 'path
+/// segments' which don't have the rest of an AST or HIR `PathSegment`.
 #[derive(Clone, Copy, Debug)]
 pub struct Segment {
     ident: Ident,
     id: Option<NodeId>,
+    /// Signals whether this `PathSegment` has generic arguments. Used to avoid providing
+    /// nonsensical suggestions.
+    has_args: bool,
```
Comment (davidtwco) [inline src/librustc_resolve/lib.rs:236]:
```suggestion
    has_generic_args: bool,
```

Might make this a little bit more clear.

**Review by davidtwco — APPROVED** [review]
LGTM, one nit, r=me when you're happy.

**Pull request merged**